File: 0645-set-mismatch/0645-set-mismatch.cpp
class Solution {
public:
    vector<int> findErrorNums(vector<int>& nums) {
        long diff = 0, sqDiff = 0;
        
        for (int i = 0; i < nums.size(); ++i){
            diff += (i + 1) - nums[i];
            sqDiff += (i + 1) * (i + 1) - nums[i] * nums[i];
        }
        
        int sum = sqDiff / diff;
        return vector<int>{(int)(sum - diff) >> 1, (int)(sum + diff) >> 1};
    }
};

File: 0645-set-mismatch/0645-set-mismatch.py
class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        # xor0 := missing, xor1: duplicate
        xor, xor0, xor1 = 0, 0, 0
        
        for n in nums:
            xor ^= n
            
        for i in range(1, len(nums) + 1):
            xor ^= i
            
        rightmostbit = xor & ~(xor - 1)
                                     
        for n in nums:
            if (n & rightmostbit) != 0:
                xor1 ^= n
            else:
                xor0 ^= n
                
        for i in range(1, len(nums) + 1):
            if (i & rightmostbit) != 0:
                xor1 ^= i
            else:
                xor0 ^= i
                
        for i in range(0, len(nums)):
            if nums[i] == xor0:
                return [xor0, xor1]
            
        return [xor1, xor0]
        